File: 02-linked-lists/20250727-intermediate/go/solution_test.go
package main

import (
	"reflect"
	"testing"
)

func TestMergeSortedLists(t *testing.T) {
	// Test case 1: Both lists empty
	list1 := &LinkedList{}
	list2 := &LinkedList{}
	result := MergeSortedLists(list1, list2)
	expected := []int{}
	if !reflect.DeepEqual(result.ToSlice(), expected) {
		t.Errorf("Test 1 failed: expected %v, got %v", expected, result.ToSlice())
	}

	// Test case 2: First list empty
	list1 = &LinkedList{}
	list2 = CreateFromSlice([]int{1, 3, 5})
	result = MergeSortedLists(list1, list2)
	expected = []int{1, 3, 5}
	if !reflect.DeepEqual(result.ToSlice(), expected) {
		t.Errorf("Test 2 failed: expected %v, got %v", expected, result.ToSlice())
	}

	// Test case 3: Second list empty
	list1 = CreateFromSlice([]int{2, 4, 6})
	list2 = &LinkedList{}
	result = MergeSortedLists(list1, list2)
	expected = []int{2, 4, 6}
	if !reflect.DeepEqual(result.ToSlice(), expected) {
		t.Errorf("Test 3 failed: expected %v, got %v", expected, result.ToSlice())
	}

	// Test case 4: Simple merge
	list1 = CreateFromSlice([]int{1, 3, 5})
	list2 = CreateFromSlice([]int{2, 4, 6})
	result = MergeSortedLists(list1, list2)
	expected = []int{1, 2, 3, 4, 5, 6}
	if !reflect.DeepEqual(result.ToSlice(), expected) {
		t.Errorf("Test 4 failed: expected %v, got %v", expected, result.ToSlice())
	}

	// Test case 5: Lists with different lengths
	list1 = CreateFromSlice([]int{1, 5, 9, 15})
	list2 = CreateFromSlice([]int{2, 3})
	result = MergeSortedLists(list1, list2)
	expected = []int{1, 2, 3, 5, 9, 15}
	if !reflect.DeepEqual(result.ToSlice(), expected) {
		t.Errorf("Test 5 failed: expected %v, got %v", expected, result.ToSlice())
	}

	// Test case 6: All elements from first list come first
	list1 = CreateFromSlice([]int{1, 2, 3, 4})
	list2 = CreateFromSlice([]int{5, 6, 7, 8})
	result = MergeSortedLists(list1, list2)
	expected = []int{1, 2, 3, 4, 5, 6, 7, 8}
	if !reflect.DeepEqual(result.ToSlice(), expected) {
		t.Errorf("Test 6 failed: expected %v, got %v", expected, result.ToSlice())
	}

	// Test case 7: All elements from second list come first
	list1 = CreateFromSlice([]int{10, 20, 30})
	list2 = CreateFromSlice([]int{1, 2, 3, 4, 5})
	result = MergeSortedLists(list1, list2)
	expected = []int{1, 2, 3, 4, 5, 10, 20, 30}
	if !reflect.DeepEqual(result.ToSlice(), expected) {
		t.Errorf("Test 7 failed: expected %v, got %v", expected, result.ToSlice())
	}

	// Test case 8: Duplicate elements
	list1 = CreateFromSlice([]int{1, 3, 5, 5})
	list2 = CreateFromSlice([]int{2, 3, 4, 5})
	result = MergeSortedLists(list1, list2)
	expected = []int{1, 2, 3, 3, 4, 5, 5, 5}
	if !reflect.DeepEqual(result.ToSlice(), expected) {
		t.Errorf("Test 8 failed: expected %v, got %v", expected, result.ToSlice())
	}

	// Test case 9: Single element lists
	list1 = CreateFromSlice([]int{5})
	list2 = CreateFromSlice([]int{3})
	result = MergeSortedLists(list1, list2)
	expected = []int{3, 5}
	if !reflect.DeepEqual(result.ToSlice(), expected) {
		t.Errorf("Test 9 failed: expected %v, got %v", expected, result.ToSlice())
	}

	// Test case 10: Negative numbers
	list1 = CreateFromSlice([]int{-5, -1, 3})
	list2 = CreateFromSlice([]int{-3, 0, 2, 4})
	result = MergeSortedLists(list1, list2)
	expected = []int{-5, -3, -1, 0, 2, 3, 4}
	if !reflect.DeepEqual(result.ToSlice(), expected) {
		t.Errorf("Test 10 failed: expected %v, got %v", expected, result.ToSlice())
	}
}

File: 02-linked-lists/20250727-intermediate/ts/solution.test.ts
import { LinkedList, mergeSortedLists } from './solution';

describe('MergeSortedLists', () => {
    test('Test 1: Both lists empty', () => {
        const list1 = new LinkedList();
        const list2 = new LinkedList();
        const result = mergeSortedLists(list1, list2);
        const expected: number[] = [];
        expect(result.toArray()).toEqual(expected);
    });

    test('Test 2: First list empty', () => {
        const list1 = new LinkedList();
        const list2 = LinkedList.createFromArray([1, 3, 5]);
        const result = mergeSortedLists(list1, list2);
        const expected = [1, 3, 5];
        expect(result.toArray()).toEqual(expected);
    });

    test('Test 3: Second list empty', () => {
        const list1 = LinkedList.createFromArray([2, 4, 6]);
        const list2 = new LinkedList();
        const result = mergeSortedLists(list1, list2);
        const expected = [2, 4, 6];
        expect(result.toArray()).toEqual(expected);
    });

    test('Test 4: Simple merge', () => {
        const list1 = LinkedList.createFromArray([1, 3, 5]);
        const list2 = LinkedList.createFromArray([2, 4, 6]);
        const result = mergeSortedLists(list1, list2);
        const expected = [1, 2, 3, 4, 5, 6];
        expect(result.toArray()).toEqual(expected);
    });

    test('Test 5: Lists with different lengths', () => {
        const list1 = LinkedList.createFromArray([1, 5, 9, 15]);
        const list2 = LinkedList.createFromArray([2, 3]);
        const result = mergeSortedLists(list1, list2);
        const expected = [1, 2, 3, 5, 9, 15];
        expect(result.toArray()).toEqual(expected);
    });

    test('Test 6: All elements from first list come first', () => {
        const list1 = LinkedList.createFromArray([1, 2, 3, 4]);
        const list2 = LinkedList.createFromArray([5, 6, 7, 8]);
        const result = mergeSortedLists(list1, list2);
        const expected = [1, 2, 3, 4, 5, 6, 7, 8];
        expect(result.toArray()).toEqual(expected);
    });

    test('Test 7: All elements from second list come first', () => {
        const list1 = LinkedList.createFromArray([10, 20, 30]);
        const list2 = LinkedList.createFromArray([1, 2, 3, 4, 5]);
        const result = mergeSortedLists(list1, list2);
        const expected = [1, 2, 3, 4, 5, 10, 20, 30];
        expect(result.toArray()).toEqual(expected);
    });

    test('Test 8: Duplicate elements', () => {
        const list1 = LinkedList.createFromArray([1, 3, 5, 5]);
        const list2 = LinkedList.createFromArray([2, 3, 4, 5]);
        const result = mergeSortedLists(list1, list2);
        const expected = [1, 2, 3, 3, 4, 5, 5, 5];
        expect(result.toArray()).toEqual(expected);
    });

    test('Test 9: Single element lists', () => {
        const list1 = LinkedList.createFromArray([5]);
        const list2 = LinkedList.createFromArray([3]);
        const result = mergeSortedLists(list1, list2);
        const expected = [3, 5];
        expect(result.toArray()).toEqual(expected);
    });

    test('Test 10: Negative numbers', () => {
        const list1 = LinkedList.createFromArray([-5, -1, 3]);
        const list2 = LinkedList.createFromArray([-3, 0, 2, 4]);
        const result = mergeSortedLists(list1, list2);
        const expected = [-5, -3, -1, 0, 2, 3, 4];
        expect(result.toArray()).toEqual(expected);
    });
});